xray/core/rolling.py

```diff
@@ -0,0 +1,262 @@
+            raise ValueError('window must be > 0')
+
+        self.obj = obj
+
+        # attributes
+        self.window = window
+        self.min_periods = min_periods
+        self.center = center
+        self.dim = dim
+
+        self._axis_num = self.obj.get_axis_num(self.dim)
+
+        self._windows = None
+        self._valid_windows = None
+        self.window_indices = None
+        self.window_labels = None
+
+        self._setup_windows(min_periods=min_periods, center=center)
+
+    def __repr__(self):
+        """provide a nice str repr of our rolling object"""
+
+        attrs = ["{k}->{v}".format(k=k, v=getattr(self, k))
+                 for k in self._attributes if getattr(self, k, None) is not None]
+        return "{klass} [{attrs}]".format(klass=self.__class__.__name__,
+                                          attrs=','.join(attrs))
+
+    @property
+    def windows(self):
+        if self._windows is None:
+            self._windows = OrderedDict(zip(self.window_labels,
+                                            self.window_indices))
+        return self._windows
+
+    def __len__(self):
+        return len(self.obj[self.dim])
+
+    def __iter__(self):
+        for (label, indices, valid) in zip(self.window_labels,
+                                           self.window_indices,
+                                           self._valid_windows):
+            if valid:
+                yield (label, self.obj.isel(**{self.dim: indices}))
+            else:
+                # return nans for this window
+                temp = self.obj.isel(**{self.dim: indices}).copy()
+                temp.values[:] = np.nan
+                yield (label, temp)
+
+    def _setup_windows(self, min_periods=None, center=False):
+        """
+        Find the indicies and labels for each window
+        """
+        self.window_labels = self.obj[self.dim]
+
+        window = int(self.window)
+        if min_periods is None:
+            min_periods = window
+
+        dim_size = self.obj[self.dim].size
+
+        stops = np.arange(dim_size) + 1
+        starts = np.maximum(stops - window, 0)
+
+        if min_periods > 1:
+            self._valid_windows = (stops - starts) >= min_periods
+        else:
+            # No invalid windows
+            self._valid_windows = np.ones(dim_size, dtype=bool)
+
+        self.window_indices = [slice(start, stop)
+                               for start, stop in zip(starts, stops)]
+
+    def _center_result(self, result):
+        """center result"""
+        shift = (-self.window // 2) + 1
+        return result.shift(**{self.dim: shift})
+
+    def sum(self):
+        """thin wrapper around bottleneck.move_sum"""
+        from .dataarray import DataArray
+
+        values = bn.move_sum(self.obj.values, window=self.window,
+                             min_count=self.min_periods, axis=self._axis_num)
+        coords = self.obj.coords
+        coords[self.dim] = self.window_labels
```

Comment:
you don't need this now because window_labels is the same as the original coordinate